marklogic-data-hub/src/main/java/com/marklogic/hub/dhs/DhsDeployer.java

```diff
@@ -55,6 +57,26 @@ public void deployAsSecurityAdmin(HubConfigImpl hubConfig) {
         dhsDeployer.deploy(hubConfig.getAppConfig());
     }
 
+    /**
+     * Compares current ML version with the minimum required version and fails immediately
+     * if current ML version is lower than the minimum required version.
+     * @param hubConfig
+     */
+    public void checkCompatibility(HubConfig hubConfig) {
+        Versions versions = new Versions(hubConfig);
+        String mlVersionString = versions.getMarkLogicVersion();
+        Versions.MarkLogicVersion mlVersion = versions.getMLVersion(mlVersionString);
+        if (!((mlVersion.getMajor() > 10) || (mlVersion.getMajor() == 10 && mlVersion.getMinor() >= 201) || (mlVersion.getMajor() == 9 && mlVersion.getMinor() >= 1100))) {
```

Comment:
We have this logic in AbstractInstallerCommand too, and it's different - because of the dependency on granular privileges, 10.0-3 or higher is needed. 

I think this question - "Given an ML version, can it support this version of DHF on DHS?" - and the question - "Given an ML version, can it support this version of DHF?" - should both be answered by the Versions.MarkLogicVersion class, which should become a real class instead of an embedded once. The problem is that the current MLV design encourages this logic to be duplicated. We should instead have methods on MLV itself that answer those questions - e.g.:

```
String mlVersion = ... // get from Manage API, or wherever, we don't care
return new MarkLogicVersion(mlVersion).supportsDataHubService();
// or .supportsDataHubFramework()
```

That class can then include all the stuff in Versions.getMLVersion. Arguably, it could go into the API project, but let's not do that yet. 

Making these changes allows you to easily test the MarkLogicVersion class too, as it's now decoupled from actually getting the ML version. So you can have a unit test that passes in all kinds of ML version strings, and MarkLogicVersion will parse it into major/minor/nightly/dateString (rename "isNightly" to "nightly" too, the Java getter will then naturally be isNightly). 

This refactoring will affect a few clients, but I think it's definitely worth doing - it gives us a much more OO design that is easier to test.

Reply:
Makes sense. I was thinking along the same lines so opened this PR as a draft.